Adds two UTxO properties

mdimjasevic opened the pull request:
This pull request implements two properties for UTxO:

* no double spending
* UTxO is outputs minuts inputs

It closes #263.

@dnadales , @uroboros: is there anything to add or change to improve this PR?

dnadales commented on byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs:
```diff
@@ -29,6 +34,44 @@ moneyIsConstant = withTests 200 . property $ do
+  let
+    utxo0 = fst . firstAndLastState $ t :: State UTXO
+    txs = traceSignals OldestFirst t :: [Signal UTXO]
+  when (all (noCommonInputs $ utxo utxo0) txs) $ do
```
Maybe we can replace `noCommonInputs` by its definition, since it makes the predicate look more similar to the specs. Also we do not need the last state it seems. So what about:

```haskell
  let
    UTxOState {utxo = utxo0} = _traceInitState t
    txs = traceSignals OldestFirst t
  when (all (\ti -> dom (txouts ti) ∩ dom utxo0 == empty) txs) 
```

mdimjasevic replied:
I moved `noCommonInputs` to the top level because both the no double spending property and the utxo is outputs minus inputs property use it. With that in mind, do you agree to leave it top-level instead of inlining it in both properties?

dnadales replied:
I would still use the definition, since it looks more similar to the specs. But I'm fine with having `noCommonInputs` if you think that's better.

Just remember to use `_traceInitState`. And I'd remove the type annotations, but again, whatever you prefer.

mdimjasevic replied:
Done! All incorporated (the definition is inlined now).

dnadales commented on byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs:
```diff
@@ -29,6 +34,44 @@ moneyIsConstant = withTests 200 . property $ do
+    utxo0 = fst . firstAndLastState $ t :: State UTXO
+    txs = traceSignals OldestFirst t :: [Signal UTXO]
+  when (all (noCommonInputs $ utxo utxo0) txs) $ do
+    (all (noCommonInputsTxs txs) (zip txs [0 .. ])) === True
```
For the remaining checks, I think it's better to use `traverse_` and `===` so that when the condition fails we can get an error message telling us which one was the transaction that was duplicated (and we can look this up in the resulting trace). So removing the redundant `do` and `ins` binding this would become:

```haskell
    traverse_ (noCommonInputsTxs txs) (zip txs [0 .. ])
   where
    noCommonInputsTxs :: MonadTest m => [Tx] -> (Tx, Int) -> m ()
    noCommonInputsTxs txs (tx, i) =
      traverse_ (\txj -> txins' txj ∩ txins' tx === empty) (take i txs)

```

Does that make sense?

mdimjasevic replied:
I like this! Incorporated it in an upcoming commit.

dnadales commented on byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs:
```diff
@@ -29,6 +34,44 @@ moneyIsConstant = withTests 200 . property $ do
+utxoDiff = property $ do
+  t <- forAll (trace @UTXO 500)
+  let
+    (utxo0, utxoSt) = (\(f, s) -> (utxo f, utxo s)) . firstAndLastState $ t :: (UTxO, UTxO)
```
You could also:

```haskell
import           Control.Arrow ((***))

    (utxo0, utxoSt) = (utxo *** utxo) . firstAndLastState $ t
```

or

```haskell
(UTxOState {utxo = utxo0}, UTxOState {utxo = utxoSt}) = firstAndLastState t
```

whatever makes sense to you.

BTW, we don't need the type annotation it seems.

mdimjasevic replied:
Thanks for the arrow hint! I added it in the upcoming commit.

When it comes to the type annotation, I know it type-checks without them; they're here to ease understanding the code (e.g., to myself in a month from now).

dnadales replied:
> When it comes to the type annotation, I know it type-checks without them; they're here to ease understanding the code (e.g., to myself in a month from now).

:+1:

dnadales commented on byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs:
```diff
@@ -29,6 +34,44 @@ moneyIsConstant = withTests 200 . property $ do
+  let
+    (utxo0, utxoSt) = (\(f, s) -> (utxo f, utxo s)) . firstAndLastState $ t :: (UTxO, UTxO)
+    txs = traceSignals OldestFirst t :: [Signal UTXO]
+  when (all (noCommonInputs utxo0) txs) $ do
```
There is a redundant `do` and a couple of redundant parentheses. You might want to run `hlint` on the file.

mdimjasevic replied:
Thanks! Fixed this in the upcoming commit.

dnadales commented on byron/ledger/executable-spec/src/Cardano/Ledger/Spec/STS/UTXO.hs:
```diff
@@ -85,3 +96,23 @@ instance STS UTXO where
+traceAddrs = mkAddr <$> [0 .. 10]
+
+-- This is mostly copy-pasted from HasTrace UTXOW
+instance HasTrace UTXO where
```
I'd be in favor of removing this instance and use `UTXOW` instead (then `txs = body <$> traceSignals OldestFirst t`). The two instances are the same modulo the witnesses, so it seems this will give use additional maintenance burden (for instance we we start adding invalid traces) at no additional benefit.

mdimjasevic replied:
That's a neat solution! Thanks, I incorporated this.

dnadales commented:
Thanks for addressing the comments Marko. Ping me when you want me to take a second look :pray:

mdimjasevic commented:
> Thanks for addressing the comments Marko. Ping me when you want me to take a second look pray

Thank you for the feedback! I'll think of coverage before pining you for another review.

uroboros commented on byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs:
```diff
@@ -29,6 +33,39 @@ moneyIsConstant = withTests 200 . property $ do
+    (utxo0, utxoSt) = (utxo *** utxo) . firstAndLastState $ t
+    txs = body <$> traceSignals OldestFirst t
+  when (all (\ti -> dom (txouts ti) ∩ dom utxo0 == empty) txs) $
+    foldl' union' empty txs ⋪ (utxo0 ∪ allTxOuts txs) === utxoSt
```
you could us Set.unions (foldl union empty) as in

  Set.unions ((fromList . txins) <$> txs)

mdimjasevic replied:
That's a fair point. I guess it boils down to what's more readable to someone and/or what is one's preference. I don't have a strong opinion here. Should I change this?

dnadales commented:
> Thank you for the feedback! I'll think of coverage before pining you for another review.

No problem.  I would address the coverage in a separate issue. Let's just close this one if possible.

mdimjasevic commented:
> No problem. I would address the coverage in a separate issue. Let's just close this one if possible.

@dnadales, I just added something so please take a look: https://github.com/input-output-hk/cardano-ledger-specs/pull/642/commits/ed339b6efc49f8c1a1884a3471dc51a9c024336a
